fs/io_zenfs.cc

```diff
@@ -361,7 +361,7 @@ IOStatus ZoneFile::Append(void* data, int data_size, int valid_size) {
   if (!active_zone_) {
     Zone* zone = zbd_->AllocateZone(lifetime_);
     if (!zone) {
-      return IOStatus::NoSpace("Zone allocation failure\n");
+      return IOStatus::NoSpace("Zone allocation failure, no active zone\n");
```

Comment:
We could also hit the "No Space" when all the zones are actually full and active zones are less than max-active-zones.  

Would it be better to dump more information such as active-zone-count, full-zone-count, total-zone-count and let the user make sense out of it ?

Reply:
This one just says zone allocation failure, nothing about how much space is left. I think caller may add the out of space message. I added this text to distinguish the error from a very similar one in line 384.

Regarding the zone stats, it is available if running with debug. In production mode I think it should be the responsibility of the caller to trigger printing of the stats, if desired.

Reply:
I think that "No active zone" adds no value to the end user. "Out of space: zone allocation failure" would be helpful.

Reply:
I see that I missed something. Out of space is already added, it is an `IOStatus::NoSpace`. No reason to put it in the message.

I made the change because identical errors were raised in multiple locations. For debugging, it was impossible to know where the error originated. We need either unique error messages, or we need to annotate the error with file name and line number.

Reply:
I think it is motivated to add "out of space" to the message as I recall people reporting "Zone allocation error" to me via email and not understanding that they actually ran out of space..

If we have different allocation errors, we should return what error we hit when returning from AllocateZone

Reply:
Printing a `IOStatus::NoSpace` error now gives something like `NoSpace: <message>`. Adding 'out of space' would print `Error ... NoSpace: Zone allocation error, no space`. Do we want that, it is sort of redundant?

> If we have different allocation errors, we should return what error we hit when returning from AllocateZone
That is a good idea, I will add that.

But we still call `AllocateZone` in multiple locations. For debugging, it is desirable to know what call site is propagating the error.